### DISCUSSION TITLE
Why is the Pico limited to 2 I2S objects?

### DISCUSSION THREAD

**Discussion by Jacob-Bishop** [micropython/micropython#14233] [RP2040 / Pico] [answered]
Hello, I'm trying to write a script to let a pi pico read audio off a mic array (and send it elsewhere for processing). The mic array I'm working with has 7 mics, so 4 data lines. As such, I ran into the limitation that the pi pico only supports two I2S objects. Two questions:

1. What's the limiting factor here? Based on a quick look at the Pico [code](https://github.com/raspberrypi/pico-extras/blob/master/src/rp2_common/pico_audio_i2s/audio_i2s.c) provided by Raspberry PI, it looks like i2s is implemented on top of the Pico's PIO, of which there seem to be 8 (but in two "blocks"). Is there some constraint that means only two of these can be used for i2s? Is it just a matter of max memory on the pico making more I2s objects not useful? Something else?

2. On a more practical note, does anyone have alternatives to suggest that would support more i2s objects? Is there some other microcontroller I should be looking at, or some add-on board that can handle more i2s lines?

Thanks!

**Comment by peterhinch** [comment]
The code is [here](https://github.com/micropython/micropython/blob/master/ports/rp2/machine_i2s.c) - I'm not sure what the limiting factor is, but you may be able to figure it out. If you can find a platform which can support seven I2S channels you need to consider the total data throughput: you may well hit another limit such as CPU cycles or bandwidth when you send it for processing - if this is done concurrently. If it is not done concurrently you need to consider storage type, size and bandwidth.

**Comment by miketeachman** [comment, marked as answer]
The decision to support 2 channels on the RP2 was made to be consistent with other MicroPython ports and consideration that this number of channels will be adequate for the vast majority of MicroPython users.  It was not based on a technical limit.  For a technical limit, the number of DMA channels on the RP2 limits the number of channels to 6.  The RP2 supports 12 DMA channels and each I2S instance uses 2 DMA channels.  There may be other technical factors which constrain the port to fewer than 6 channels.

**Comment by Jacob-Bishop** [comment]
Thanks all for the guidance, and the speedy response! I'll do some experimentation...
